ipa-core/src/query/runner/oprf_ipa.rs

```diff
@@ -31,12 +31,12 @@ use crate::{
 
 pub struct OprfIpaQuery<'a, HV> {
     config: IpaQueryConfig,
-    key_registry: Arc<KeyRegistry<KeyPair>>,
+    key_registry: Arc<KeyRegistry<PrivateKeyOnly>>,
     phantom_data: PhantomData<&'a HV>,
 }
 
 impl<'a, HV> OprfIpaQuery<'a, HV> {
-    pub fn new(config: IpaQueryConfig, key_registry: Arc<KeyRegistry<KeyPair>>) -> Self {
+    pub fn new(config: IpaQueryConfig, key_registry: Arc<KeyRegistry<PrivateKeyOnly>>) -> Self {
```

Comment:
can you try `Arc<dyn PrivateKeyRegistry>` here? I think in this case you wouldn't require conversion in unit tests

Reply:
This gives me the following compilation error. 

```
error[E0277]: the size for values of type `dyn hpke::registry::PrivateKeyRegistry` cannot be known at compilation time
   --> ipa-core/src/query/runner/oprf_ipa.rs:84:38
    |
84  | ...                   .decrypt(key_registry.as_ref())
    |                        ------- ^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |                        |
    |                        required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `dyn hpke::registry::PrivateKeyRegistry`
note: required by a bound in `report::EncryptedOprfReport::<BK, TV, TS, B>::decrypt`
   --> ipa-core/src/report.rs:313:24
    |
311 |     pub fn decrypt(
    |            ------- required by a bound in this associated function
312 |         &self,
313 |         key_registry: &impl PrivateKeyRegistry,
    |                        ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `EncryptedOprfReport::<BK, TV, TS, B>::decrypt`
help: consider removing this method call, as the receiver has type `std::sync::Arc<(dyn hpke::registry::PrivateKeyRegistry + 'static)>` and `std::sync::Arc<(dyn hpke::registry::PrivateKeyRegistry + 'static)>: Sized` trivially holds
    |
84  -                             .decrypt(key_registry.as_ref())
84  +                             .decrypt(key_registry)
    |
help: consider relaxing the implicit `Sized` restriction
   --> ipa-core/src/report.rs:313:47
    |
313 |         key_registry: &impl PrivateKeyRegistry + ?Sized,
    |                                                ++++++++

warning: `ipa-core` (lib test) generated 1 warning (1 duplicate)
error: could not compile `ipa-core` (lib test) due to 1 previous error; 1 warning emitted
```

Reply:
The suggestion 
```
84  -                             .decrypt(key_registry.as_ref())
84  +                             .decrypt(key_registry)
``` 
produces an incompatible type, as decrypt expects a reference. If we borrow here, we then get:
```
the trait `PrivateKeyRegistry` is not implemented for `Arc<dyn PrivateKeyRegistry>`
```

Adding `?Sized` to in `ipa-core/src/report.rs` simply leads to errors elsewhere about the size being unknown at compile times. 

It seems like keeping the conversion might be the most straight forward path, but happy to be pointed in a different direction.

Reply:
pushed the change that can make `OprfIpa` accept `Arc<dyn>`. Again, I realized how clunky dynamic dispatch is in Rust and we should probably just parametrize `OprfIpa` with another generic

Reply:
ty! that works. now let me try to remove the implementation of `From`.